pkg/persistence/disk_persistence.go

```diff
@@ -65,6 +82,53 @@ func (ds *diskPersistence) Save(data []byte, dirName, fileName string) error {
 	return write(fmt.Sprintf("%s/%s/%s", dirPath, dirName, fileName), data)
 }
 
+func (ds *diskPersistence) Snapshot(data []byte, dirName, fileName string) error {
+	ds.snapshotMutex.Lock()
+	defer ds.snapshotMutex.Unlock()
+
+	if len(dirName) > maxFileNameLength {
+		return fmt.Errorf(
+			"the maximum directory name length of [%v] exceeded for [%v]",
+			maxFileNameLength,
+			dirName,
+		)
+	}
+
+	snapshotSuffix := ds.snapshotSuffixGenerator()
+
+	maxSnapshotFileNameLength := maxFileNameLength - len(snapshotSuffix)
+	if len(fileName) > maxSnapshotFileNameLength {
+		return fmt.Errorf(
+			"the maximum file name length of [%v] exceeded for [%v]",
+			maxSnapshotFileNameLength,
+			fileName,
+		)
+	}
+
+	dirPath := fmt.Sprintf("%s/%s", ds.dataDir, snapshotDir)
+	err := createDir(dirPath, dirName)
+	if err != nil {
+		return err
+	}
+
+	filePath := fmt.Sprintf("%s/%s/%s", dirPath, dirName, fileName+snapshotSuffix)
+
+	// very unlikely but better fail than overwrite an existing file
+	if canWrite := isNotExist(filePath); !canWrite {
```

Comment:
This `if` statement feels like it's giving us more code just to invert a condition that can already be articulated. Why not `if fileAlreadyExists(filePath) { ...`?

Reply:
Addressed in the follow-up PR: #50 

I've got rid of the `canWrite` variable and renamed the `isNotExist` method a bit. But, I think we should stick to the current condition format because we can save the snapshot only if we have the `oserror.ErrNotExist` error returned from the check. This is easier to check than checking file existence because in case of an existing file the `os.Stat` method can return not just a nil-error but also several other errors types, for example `oserror.ErrPermission`.

Reply:
Heh, I remember it took me a while to understand how `os.Stat` when I was first reviewing this code.